### PR TITLE
(PUP-8687) Add support for logging with JSON lines output format

### PR DESCRIPTION
**Pull request by smortex**
Puppet had logging to JSON for some time, but the generated log file is not a valid JSON document because of the way a log file is written to.  [JSON lines] (aka newline-delimited JSON) format is a text file format that use newlines to separate JSON documents, making it easy to process entries one at a time, and therefore not suffering from this deficiency.

Convention suggests that [JSON lines] files may be saved with the file extension `.jsonl`.

[JSON lines]: http://jsonlines.org/

### CODE REVIEW

File: lib/puppet/util/log/destinations.rb [hunk 1]
```diff
@@ -111,6 +112,8 @@ def handle(msg)
     if @json > 0
       @json > 1 ? @file.puts(',') : @json = 2
       @file.puts(Puppet::Util::Json.dump(msg.to_structured_hash))
+    elsif @jsonl
+      @file.puts(Puppet::Util::Json.dump(msg.to_structured_hash))
```
Comment (joshcooper) [inline lib/puppet/util/log/destinations.rb]: Could you add a unit test, something like https://github.com/puppetlabs/puppet/blob/master/spec/unit/util/log/destinations_spec.rb#L82-L94, but with a file with `.jsonl` extension?

Comment (smortex) [reply lib/puppet/util/log/destinations.rb]: D'oh, I failed to find the JSON test.  I just added a commit to check for JSON lines as well.  Thanks for the pointer to the test!

**Comment by joshcooper** [comment]
LGTM. One nit, could you squash the code and test changes into one commit and force push to your branch? Thanks!

**Comment by smortex** [comment]
@joshcooper, done and rebased on top of master :wink:, thanks!

**Pull request merged**